ykrt/src/compile/jitc_yk/codegen/x64/mod.rs

```diff
@@ -2878,6 +2874,58 @@ impl CompiledTrace for X64CompiledTrace {
         &self.deoptinfo[&usize::from(gidx)].guard
     }
 
+    /// Patch the address of a side-trace directly into the parent trace.
+    /// `gidx`: The guard to be patched.
+    /// `staddr`: The address of the side-trace.
+    fn patch_guard(&self, gidx: GuardIdx, staddr: *const std::ffi::c_void) {
+        // Make the trace writable while keeping it executable.
+        // FIXME: This might not work on other architectures or operating systems. However, the
+        // solution is very involved and requires changing `mt.rs` to make sure no other thread is
+        // accessing the trace (which includes executing the trace itself or any of it's
+        // side-traces, or side-tracig a guard failure).
+        let trace_addr = self.buf.ptr(AssemblyOffset(0)) as *mut libc::c_void;
+        let res = unsafe {
+            libc::mprotect(
+                trace_addr,
+                self.buf.len() as libc::size_t,
+                libc::PROT_WRITE | libc::PROT_EXEC,
```

Comment:
Asides from the concurrency issues mentioned above, we also shouldn't allocate pages W+X. Can we put a FIXME to remind us?

Reply:
Also, do we ever reset the page permissions?

Reply:
It's fine for us to *temporarily* mark pages as such: we just don't want to *leave* them in this state. 

However the particular way we're doing it here is a race condition. If thread1 `mprotect`s loc X to WRITE + EXEC *and* thread2 does the same, then thread1 can `mprotect` X back to EXEC before thread2 has had the chance to do its update. We need to use a lock *somewhere* to prevent this race condition.

Fortunately, because we're calling this from a parallel compile job in `mt`, we can do something pretty dumb at first: a global `MPROTECT` lock which we hold just before the first `mprotect` call and which we release immediately after the second `mprotect` call.

Reply:
Doesn't this fixme already cover this? `FIXME: This might not work on other architectures or operating systems.`

Reply:
The race condition is not architecture dependent.

Reply:
> However the particular way we're doing it here is a race condition.

Ah, I see it now too. Just a simple spin lock is fine, right?

Reply:
We need to hold e.g. a `Mutex`. Whether it's a spin lock or not shouldn't concern us.

Reply:
Yeah, I meant a lock that waits until it gets access.

Reply:
I'm not sure if [3897c38](https://github.com/ykjit/yk/pull/1591/commits/3897c38c05bfe7aae1765d93ec4ade1aba7eddbe) is what you had in mind.

Reply:
More-or-less. I would do `Mutex::new(());` i.e. use the unit type rather than an integer. Conventionally in yk we also spell "lock" variables `lk`. Those are very minor, cosmetic changes, so we might as well do 'em, but they're a 2 second job.

One thing that I am chewing on right now is whether we should `panic` if either `mprotect` call fails. *Mostly* these can only fail if *we* do something stupid *but* if user code calls `mlockall` then `mprotect` would fail. Ideally we would throw a warning to `stderr` with `Verbosity::Error` or maybe `Verbosity::Warning` but that would mean we'd need a fallback mechanism for when patching fails. I suspect this commit removes the fallback mechanism, at which point the `panic` is OK until/if we ever see this happen in the wild.

Reply:
> It's fine for us to temporarily mark pages as such

Well, it depends. There are configurations, even on linux, where a call to mprotect with W+X would fail.

Granted, that's off by default on linux, so we are unlikely to encounter it...

Reply:
This is a problem for JITs in general and we're not going to solve it in this PR, nor do I think we need worry about it.

Reply:
> I would do Mutex::new(()); i.e. use the unit type rather than an integer.

I'm using the parking_lot mutex because we already import that. And it needs a u32. Want me to witch to `std::sync` in instead?

Reply:
Addressed comments in [f5934a6](https://github.com/ykjit/yk/pull/1591/commits/f5934a6e8de19c96000a48aa20c1dbc4bd96519f) [872f241](https://github.com/ykjit/yk/pull/1591/commits/872f241bfed964aa0a005d4d8cdbe96a6c078351) [07f445d](https://github.com/ykjit/yk/pull/1591/commits/07f445d465ba3f6df81bdf44dc0fde4243388c31)